### ISSUE TITLE
Potential performance regression in neo4j 5.x with bolt, maybe linked to buffering 

### ISSUE THREAD

**Issue by kristofka** [neo4j/neo4j#13045]
## Report

While working on a third party bolt driver, I've realised that queries made against neo4j 5.x were consistently slower than those made against neo4j 4.x (or 3.5 for that matter).
Thinking the issue might be with my driver, or with some configuration issue, I've tried the same things with the browser and the official python driver with similar result.

### Steps to reproduce
1. Using docker images 4.4 and 5.4 with bolt mapped on localhost:7687 and localhost:7689 respectively.
2. Create 100K nodes on both in cypher-shell by doing `unwind range(1,100000) as n create (:Test {value: n});|`
3. with the following python code :
```python
class Bencher:
     def __init__(self, uri, user, password):
            self.driver = GraphDatabase.driver(uri, auth=(user, password))
     def close(self):
            self.driver.close()
     def print_times(self):
            with self.driver.session() as session:
            start = time.time()
            res = session.execute_read(self._yield_and_return)
            end = time.time()
            print((end - start) * 1000)
    @staticmethod
    def _yield_and_return(tx):
        result = tx.run("match (n) return n limit(100000)")
        return result.to_eager_result()
```
5. run `neo4 = Bencher("bolt://localhost:7687", "neo4j", "testtest")` and `neo5 = Bencher("bolt://localhost:7689", "neo4j", "testtest")`
6. run `neo4.print_times()` and `neo5.print_times`

### Expected behavior
Times should be roughly equal with both.

### Actual behavior
With neo4j python driver on my machine neo4j 4.4 gives on average 6 seconds and neo4j 5.x gives 10.5 seconds

### Behaviour with other drivers

In the neo4j browser, the same query seems much more sluggish with version 5.x than version 4.4 although I haven't measured it reliably.

With the driver I'm currently working on (I'm not ready to share the code yet), the same query takes, on average 370ms with neo4j 4.4 and 4+ seconds with neo4j 5.x

![visualization](https://user-images.githubusercontent.com/140354/219503909-a3a62310-9137-42a8-a088-45d8383a3c84.png)

### Notes

After investigating those queries with wireshark, I've noted that neo4j 5.x sends roughly 100 times as many packets as neo4j 4.x / 3.5, packets length statistics are reproduced below :

- neo4j 4.4

|level|parent         |Topic / Item    |Count|Average |Min Val|Max Val|Rate (ms)|Percent|Burst Rate|Burst Start|
|-----|---------------|----------------|-----|--------|-------|-------|---------|-------|----------|-----------|
|0    |               |Packet Lengths  |400  |7727.54 |44     |16388  |0.1707   |100%   |2.2100    |2.366      |
|1    |/Packet Lengths|0-19            |0    |-       |-      |-      |0.0000   |0.00%  |-         |-          |
|1    |/Packet Lengths|20-39           |0    |-       |-      |-      |0.0000   |0.00%  |-         |-          |
|1    |/Packet Lengths|40-79           |86   |55.98   |44     |66     |0.0367   |21.50% |0.3400    |2.446      |
|1    |/Packet Lengths|80-159          |6    |118.17  |82     |152    |0.0026   |1.50%  |0.0600    |2.363      |
|1    |/Packet Lengths|160-319         |0    |-       |-      |-      |0.0000   |0.00%  |-         |-          |
|1    |/Packet Lengths|320-639         |1    |524.00  |524    |524    |0.0004   |0.25%  |0.0100    |2.543      |
|1    |/Packet Lengths|640-1279        |1    |888.00  |888    |888    |0.0004   |0.25%  |0.0100    |2.642      |
|1    |/Packet Lengths|1280-2559       |0    |-       |-      |-      |0.0000   |0.00%  |-         |-          |
|1    |/Packet Lengths|2560-5119       |0    |-       |-      |-      |0.0000   |0.00%  |-         |-          |
|1    |/Packet Lengths|5120 and greater|306  |10078.69|8164   |16388  |0.1306   |76.50% |1.8800    |2.390      |

- neo4j 5.x

|level|parent         |Topic / Item    |Count|Average |Min Val|Max Val|Rate (ms)|Percent|Burst Rate|Burst Start|
|-----|---------------|----------------|-----|--------|-------|-------|---------|-------|----------|-----------|
|0    |               |Packet Lengths  |31538|299.32  |44     |16388  |5.2805   |100%   |16.5300   |6.490      |
|1    |/Packet Lengths|0-19            |0    |-       |-      |-      |0.0000   |0.00%  |-         |-          |
|1    |/Packet Lengths|20-39           |0    |-       |-      |-      |0.0000   |0.00%  |-         |-          |
|1    |/Packet Lengths|40-79           |4    |50.00   |44     |56     |0.0007   |0.01%  |0.0200    |3.284      |
|1    |/Packet Lengths|80-159          |15050|132.74  |82     |134    |2.5199   |47.72% |8.7500    |6.490      |
|1    |/Packet Lengths|160-319         |9732 |234.87  |160    |290    |1.6295   |30.86% |5.2600    |4.880      |
|1    |/Packet Lengths|320-639         |5201 |437.62  |324    |632    |0.8708   |16.49% |4.0000    |5.125      |
|1    |/Packet Lengths|640-1279        |1303 |820.02  |640    |1240   |0.2182   |4.13%  |0.8500    |3.675      |
|1    |/Packet Lengths|1280-2559       |134  |1562.68 |1297   |2474   |0.0224   |0.42%  |0.1500    |5.725      |
|1    |/Packet Lengths|2560-5119       |13   |3592.46 |2630   |4814   |0.0022   |0.04%  |0.0400    |3.803      |
|1    |/Packet Lengths|5120 and greater|101  |15400.06|5360   |16388  |0.0169   |0.32%  |0.9400    |4.407      |

I believe there is an issue here, it might not be the number of packets sent, but it clearly affects communications through bolt.

**Comment by kristofka** [comment]
Additionnal data point, java driver : 

code : 
```java
package org.example;

import org.neo4j.driver.AuthTokens;
import org.neo4j.driver.Driver;
import org.neo4j.driver.GraphDatabase;
import org.neo4j.driver.Query;

import static org.neo4j.driver.Values.parameters;

public class Main implements AutoCloseable {
    private final Driver driver;

    public Main(String uri, String user, String password) {
        driver = GraphDatabase.driver(uri, AuthTokens.basic(user, password));
    }

    @Override
    public void close() throws RuntimeException {
        driver.close();
    }

    public void printGreeting(final String message) {
        long start = System.currentTimeMillis();
        try (var session = driver.session()) {
            var greeting = session.executeWrite(tx -> {
                var query = new Query("match (n) return n", parameters("message", message));
                var result = tx.run(query);
                return result.stream().toList();
            });
            long elapsedTimeMillis = System.currentTimeMillis()-start;
            System.out.println(elapsedTimeMillis);
        }
    }

    public static void main(String... args) {
        try (var greeter = new Main("bolt://localhost:7687", "neo4j", "testtest")) {
            greeter.printGreeting("hello, world");
        }
    }
}
```

With neo4j 4.4 average is around 1s, with neo4j 5.x it's around 5.5 secs. So even with the official java driver, there's a 5+x slowdown. 
I strongly believe this is an issue with the bolt server. I also believe ALL drivers are affected, in some fashion.

**Comment by gjmwoods** [comment]
Hi @kristofka, many thanks for this issue and for all the detail provided on it.

You've correctly identified that a change in buffering has lead to a performance regression. A refactor we implemented in 5.0 had the unintended consequence that each record written to the buffer would be flushed immediately where as pre 5.0 we would wait for the internal buffer to fill before flushing.

We noticed this in our own internal benchmarking but decided its impact on real world use cases would be limited. Small records such as in your benchmarks will show the significant performance degradation as you have seen but typical records sent over the wire are likely much larger and therefore the regression less noticeable.

We do have a fix for it that will be release in 5.6 in roughly a months time. Please note that as part of the refactor we discovered a thread safety issue which means we do expect to see some regression.

Many thanks again for this gh issue.

**Comment by kristofka** [comment]
Hi @gjmwoods , thank you for your comment.

Nice to hear that the buffering issue will be fixed in 5.6 . In the mean time, the issue is less serious than I initially thought. As you said, the problem pretty much disappears with larger records (meaning I can live with the really small performance degradation in neo4j 5.x since it's pretty much negligible), I didn't think to benchmark with larger records.

Before 5.6 lands, it's pretty easy to workaround the issue by batching the records with cypher when the query is expected to return many small records (eg, assuming apoc is installed the previous query : 

`match (n) return n` could be turned into `match (n:Test) with apoc.coll.partition( collect(n), 100) as p unwind p as res return res`, while slightly uglier, the small change allows to have performances that are in line with neo4j 4.4 - and I can live with that ).

With those information I have rerun the benchmarks from the driver I'm working on (still with the same 100k nodes in the db) : 
 - bolt v4.4 is the query `match (n:Test) return n` run against neo4j 4.4 ;
 - bolt v4.4 collected is the query `match (n:Test) with apoc.coll.partition( collect(n), 100) as p unwind p as res return res` run against neo4j 4.4;
 -  bolt v5.x collected is the query `match (n:Test) with apoc.coll.partition( collect(n), 100) as p unwind p as res return res` run against neo4j 5.4 ;

![bench_collected](https://user-images.githubusercontent.com/140354/220677104-d26015b5-9d7b-493c-a4d9-7381e9770484.png)

details : 
```
##### With input g - 100000 recs #####
Name                             ips        average  deviation         median         99th %
c: bolt v4.4 collected          3.84      260.59 ms    ±12.24%      268.53 ms      333.78 ms
e: bolt v5.x collected          3.14      318.19 ms     ±7.64%      320.23 ms      391.27 ms
c: bolt v4.4                    2.60      384.65 ms    ±12.30%      375.05 ms      533.26 ms

Comparison: 
c: bolt v4.4 collected          3.84
e: bolt v5.x collected          3.14 - 1.22x slower +57.60 ms
c: bolt v4.4                    2.60 - 1.48x slower +124.07 ms
```

I'm fine with those results. They might suggest that buffering could perhaps be improved, even in v4, but I don't know if it'd be worth it.

Thanks again for the reply and future fix.
